src/osd/osd_types.cc

```diff
@@ -3852,14 +3864,16 @@ struct compact_interval_t {
     decode(acting, bl);
     DECODE_FINISH(bl);
   }
+  std::string fmt_print() const {
+    return fmt::format("([{},{}] acting={})", first, last, acting);
```

Comment:
if save the efforts to write the boilerplate code for specializing a `fmt::formatter<foobar>`. but the convenience is archived at the expense of performance.

if we care about performance when formatting something, we should avoid this. it always creates a temporary string and then drops it on the floor after deep copying it to

* the buffer provided by fmtlib or 
* whatever buffer passed to `fmt::format_to()` or
* yet another a `std::string` when it comes to `fmt::to_string()`

Reply:
@tchaikov 

Take a look at an example benchmark I run:
The code is in https://godbolt.org/z/dE95K6aM3

With fmtlib10, gcc 12, -O3:
For vl=10, which creates a sensible-sized output:

- the indirect use of the fmt formatter adds up to 10% to the execution time (dropping to ~3 for larger vl); But:

- using ostream is 3 times slower.

We are, in many cases, already using a 'print' function to either:
- help with formatting (both for ostream and for fmt formatters) of derived classes, or
- to code the formatting once for both methods.

What I am trying to do here is to nudge the developers in the right direction: to improve performance
by using 'FMT first'. I'd suggest that, while very slightly less than the absolute optimum, is a step in the
right direction. Note that I am mainly replacing the 'fmt/ostream' implementations.

Reply:
> to code the formatting once for both methods.

this is a strong constraint of the design . i think the need of fmtlib formatter comes from crimson, where we use the Seastar's logging subsystem, which in turn uses fmtlib for formatting. on in the meantime, Ceph's native logging subsystem is based on `std::ostream` and `operator<<`. 

despite that fmtlib is able to format into a ostream. but to use it, we have two solutions:

1. to change the way how we write logging messages -- to have yet another wrapper around `MutableEntry`. and 
2. to create a templated `operator<<` which prints to dout using fmtlib.

> ```
> <stdin>:374:1: no absolute or relative includes please
> ```

the code does not compile. neither am i in favor of `fmt::ostream_formatter` solution, it creates a temporary `ostream` from a streambuf, inserts the stuff to be formatted, and then throw it away. seems it is worse than the combo of temporary `fmt::format_context` and `std::string`, per your testing. but what's wrong with inheriting from a, for instance `fmt::formatter<std::string_view>` and implement `format(const foobar&, fmt::format_context&)`? `foobar` can implement `auto format(fmt::format_context& ctx) const -> decltype(ctx.out())`, if developer don't really want to specialize a `fmt::formatter`. i am not claiming that this solution is better than the one proposed in this pull request. i just failed to find the constraits of the solution, and the comparisons of the performance benchmarks and the pros and cons of different approaches in the commit message. it seems that the commit message implies that this is the only (obvious) way to address the problem. but IMHO, there are more than one way to do this.

Reply:
@tchaikov : Thanks for the review. A few general comments:

1 - the need for fmt formatters is from the Classic side, too. I am trying to make it the default everywhere. (and yes - I agree with what you say: at some point we should probably switch to using the Crimson logger format).

2 - I do not dispute the suggestion that usually the best approach is to create a specific formatter. I am suggesting a (might be temporary) shortcut.

3 - I agree - my suggestion is not the only possible one. It's just convenient when we have a templated hierarchy (and we already use similar mechanisms for dump() etc on such hierarchies.

4 - I'll go over the PR, to make sure I implement the operator<< directly from the formatter wherever convenient.

and

5 - the code compiles for me, both for classic & crimson, gcc & clang. I will verify that again now.

Reply:
> @tchaikov : Thanks for the review. A few general comments:
> 
> 1 - the need for fmt formatters is from the Classic side, too. I am trying to make it the default everywhere. (and yes - I agree with what you say: at some point we should probably switch to using the Crimson logger format).

sorry for the confusion. i didn't claim that we should switch to crimson's logging format. it turns out to be non-trivial in the term of the amount of work, and we have much more important work to do in Ceph. i was just thinking aloud when trying to understand the difficultly of using fmtlib in the existing logging subsystem.

> 
> 2 - I do not dispute the suggestion that usually the best approach is to create a specific formatter. I am suggesting a (might be temporary) shortcut.

IMHO, a reviewer not only evaluates if a change is an improvement, he/she also should explore the possible solutions if there is any, and encourage the contributor to evaluate them if he/she hasn't done so. i think in the longer run, the change proposed by your change will be superseded once `std::format()` is around so i'd avoid reworking on this part.

> 
> 3 - I agree - my suggestion is not the only possible one. It's just convenient when we have a templated hierarchy (and we already use similar mechanisms for dump() etc on such hierarchies.
> 
> 4 - I'll go over the PR, to make sure I implement the operator<< directly from the formatter wherever convenient.
> 
> and
> 
> 5 - the code compiles for me, both for classic & crimson, gcc & clang. I will verify that again now.

since you shared a link to godbolt. it's does not compile in godbolt. probably i am missing something obvious?